### PR TITLE
Helm improvement

### PR DESCRIPTION
**Pull request by tunacicek**
## Description
helm improvement.
* Add _helper.tpl and define functions for (Chart Name, etc.)

### CODE REVIEW

File: charts/registry/templates/keycloak/database-credentials.yaml [hunk 1]
```diff
@@ -19,10 +19,12 @@
 # SPDX-License-Identifier: Apache-2.0
 ###############################################################
 
+{{- $fullName := include "chart.fullname" . }}
+{{- $keycloak_sec_name := printf "%s-keycloak-sec" $fullName }}
 apiVersion: v1
 kind: Secret
 metadata:
-  name: keycloak-database-credentials
+  name: {{ $keycloak_sec_name }}
```
Comment (florianrusch-zf) [inline charts/registry/templates/keycloak/database-credentials.yaml:27]: why not simply like this? Of course, this could be also applied to the other resources as well ;-) 
```suggestion
apiVersion: v1
kind: Secret
metadata:
  name: {{ include "chart.fullname . }}-keycloak-sec
```

File: charts/registry/values.yaml [hunk 2]
```diff
@@ -18,14 +18,16 @@
...
+nameOverride:
+fullnameOverride:
 # enables the default postgres database
 enablePostgres: true
 # enables the default keycloak identity provider
 # relies on a postgres instance
 enableKeycloak: false
 
 registry:
+  name: registry
```
Comment (florianrusch-zf) [inline charts/registry/values.yaml:30]: Whats the purpose of this? Couldn't find any usage of this.

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

File: charts/registry/templates/keycloak/database-credentials.yaml (hunk 1 above)
Comment (tunacicek) [reply charts/registry/templates/keycloak/database-credentials.yaml]: Thanks for the hint :) 
Files adjusted.

File: charts/registry/values.yaml (hunk 2 above)
Comment (tunacicek) [reply charts/registry/values.yaml]: You are right. The property is not used. 
I removed this property.

File: charts/registry/templates/registry/registry-ingress.yaml [hunk 3]
```diff
@@ -19,17 +19,14 @@
...
 ###############################################################
 
-{{- $deployment_name := printf "cx-%s-registry" .Release.Name }}
-{{- $svc_name := printf "%s-svc" $deployment_name }}
-{{- $ingr_name := printf "%s-ingr" $deployment_name }}
 apiVersion: networking.k8s.io/v1
 kind: Ingress
 metadata:
-  name: {{ $ingr_name }}
+  name: {{ include "chart.fullname" . }}-ingr
```
Comment (florianrusch-zf) [inline charts/registry/templates/registry/registry-ingress.yaml:25]: The suffix is not necessary, as you just create one ingress with your helm chart ;-) 

```suggestion
  name: {{ include "chart.fullname" . }}
```

File: charts/registry/templates/keycloak/database-credentials.yaml [hunk 4]
```diff
@@ -22,7 +22,7 @@
 apiVersion: v1
 kind: Secret
 metadata:
-  name: keycloak-database-credentials
+  name: {{ include "chart.fullname" . }}-keycloak-sec
```
Comment (florianrusch-zf) [inline charts/registry/templates/keycloak/database-credentials.yaml:25]: I would choose a more descriptive name. E.g.:
```suggestion
  name: {{ include "chart.fullname" . }}-keycloak-db-creds
```

If you change this, please update also the references to this secret (which I couldn't find 🤔 - how is this secret been used?).

File: charts/registry/templates/registry/registry-ingress.yaml (hunk 3 above)
Comment (florianrusch-zf) [reply charts/registry/templates/registry/registry-ingress.yaml:25]: This could be also applied to all the other resources (except the keycloak stuff 😉)

File: charts/registry/templates/registry/registry-service.yaml [hunk 5]
```diff
@@ -18,18 +18,16 @@
...
-    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+    {{- include "chart.labels" . | nindent 4 }}
 spec:
   type: {{ .Values.registry.service.type }}
   ports:
     - port: {{ .Values.registry.service.port }}
       targetPort: {{ .Values.registry.containerPort }}
   selector:
-    app: {{ $deployment_name }}
+    {{- include "chart.selectorLabels" . | nindent 6 }}
```
Comment (florianrusch-zf) [inline charts/registry/templates/registry/registry-service.yaml:33]:
```suggestion
    {{- include "chart.selectorLabels" . | nindent 4 }}
```

File: charts/registry/templates/tests/test-connection.yaml [hunk 6]
```diff
@@ -17,10 +17,15 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 ###############################################################
+
+{{- $fullName := include "chart.fullname" . }}
+{{- $testconnection_name := printf "%s-test-script-pod" $fullName }}
+{{- $testsec_name := printf "%s-test-credentials" $fullName }}
+{{- $testconfig_name := printf "%s-test-script" $fullName }}
```
Comment (florianrusch-zf) [inline charts/registry/templates/tests/test-connection.yaml:24]: I would handle that the same way as in the other templates. But it's up to you if you want to keep it consistent or not 😉

File: charts/registry/templates/tests/test-credentials.yaml [hunk 7]
```diff
@@ -17,10 +17,14 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 ###############################################################
+
+{{- $fullName := include "chart.fullname" . }}
+{{- $testsec_name := printf "%s-test-credentials" $fullName }}
+{{- $svc_name := printf "%s-svc" $fullName }}
```
Comment (florianrusch-zf) [inline charts/registry/templates/tests/test-credentials.yaml:23]: Same as above in `test-connection.yaml`

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

File: charts/registry/templates/_helpers.tpl [hunk 8]
```diff
@@ -0,0 +1,48 @@
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "chart.name" -}}
```
Comment (florianrusch-zf) [inline charts/registry/templates/_helpers.tpl:4]: To be honest, I'm not sure, if using `chart` is a good idea or not. Because if I remember correctly, all "define" methods are registered in the same context. This means that if someone includes the DTR helm chart in another helm chart as a dependency (as we plan to do with the umbrella helm chart), we might run into problems if another helm chart defines the same methods. But I'm really not sure about that.

@evegufy do you have more experience with this? Otherwise I would just do a quick test on my local computer and test that.

Comment (evegufy) [reply charts/registry/templates/_helpers.tpl:4]: @florianrusch-zf My guess is that it remains in the context of the subchart even if you add it as dependency to another chart. Still, it's good to test it on your local.

Comment (evegufy) [reply charts/registry/templates/_helpers.tpl:4]: @florianrusch-zf @tunacicek even tough it may not lead to immediate issues, I think it's a good practice to define all templates in a unique manner, 'dtr.name' for instance would be better. And by doing so, we may also save some troubleshooting time in the future for use case like an umbrella chart.

File: charts/registry/templates/keycloak/database-credentials.yaml [hunk 9]
```diff
@@ -22,7 +22,7 @@
 apiVersion: v1
```
Comment (florianrusch-zf) [inline charts/registry/templates/keycloak/database-credentials.yaml:22]: Regarding the postgres creds I would recommend to make it also possible, to use an already existing secret. Would it make sense to address this also in this PR, or should it be better in a separate PR?

Comment (tunacicek) [reply charts/registry/templates/keycloak/database-credentials.yaml:22]: I will create a new Issue for this topic :) 
Thanks for the feedback.

File: charts/registry/templates/_helpers.tpl (hunk 8 above)
Comment (tunacicek) [reply charts/registry/templates/_helpers.tpl:4]: @evegufy good idea. Thanks :) 
I adjust the templates.

**Review by evegufy — CHANGES_REQUESTED** [review]
Hi @tunacicek, I think it was quite good how you used the templates before this [commit](https://github.com/eclipse-tractusx/sldt-digital-twin-registry/pull/281/commits/a8ec9a6bd7b9cc564bab24c3a63e021fa99c0b9b).
With my comment regarding defining the templates in unique manner, I meant instead of defining chart.name it's better to dtr.name and so on, here an [example](https://github.com/eclipse-tractusx/policy-hub/blob/main/charts/policy-hub/templates/_helpers.tpl#L4).
Sorry if the [commit](https://github.com/eclipse-tractusx/sldt-digital-twin-registry/pull/281/commits/a8ec9a6bd7b9cc564bab24c3a63e021fa99c0b9b) was inspired by the portal chart, I put unique template names also on my TODO list.

**Comment by tunacicek** [comment]
@evegufy : I got it. I've made the changes, thanks.

**Comment by evegufy** [comment]
> @evegufy : I got it. I've made the changes, thanks.

@tunacicek the templates look good to me now :+1:

Probably not related to the templates, but when installing locally (without any additional configuration) the readiness probes for the deployment are failing:
image
Other than the readiness probe, everything looks fine (pod starts up), at least as far I can tell.

File: charts/registry/templates/keycloak/init-script-configmap.yaml [hunk 10]
```diff
@@ -20,7 +20,7 @@
 kind: ConfigMap
 apiVersion: v1
 metadata:
-  name: {{ include "chart.fullname" . }}-keycloak-configmap
+  name: {{ include "dtr.fullname" .  }}-keycloak-configmap
```
Comment (florianrusch-zf) [inline charts/registry/templates/keycloak/init-script-configmap.yaml:23]:
```suggestion
  name: {{ include "dtr.fullname" . }}-keycloak-configmap
```

File: charts/registry/templates/registry/registry-deployment.yaml [hunk 11]
```diff
@@ -21,23 +21,23 @@
...
     metadata:
       labels:
-        {{- include "chart.selectorLabels" . | nindent 8 }}
+        {{- include "dtr.selectorLabels" . | nindent 8 }}
     spec:
       securityContext:
         runAsUser: 100
       containers:
-        - name: {{ .Chart.Name }}
+        - name: {{ include "dtr.fullname" . }}
```
Comment (florianrusch-zf) [inline charts/registry/templates/registry/registry-deployment.yaml:40]: Whats the reason why you changed it from `{{ .Chart.Name }}` to the fullname?
Chart name was already fine :)

File: charts/registry/templates/registry/registry-service.yaml [hunk 12]
```diff
@@ -21,13 +21,13 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ include "chart.fullname" . }}
+  name: {{ include "dtr.fullname" .  }}
```
Comment (florianrusch-zf) [inline charts/registry/templates/registry/registry-service.yaml:24]:
```suggestion
  name: {{ include "dtr.fullname" . }}
```

**Review by florianrusch-zf — CHANGES_REQUESTED** [review]

File: charts/registry/templates/registry/registry-deployment.yaml (hunk 11 above)
Comment (tunacicek) [reply charts/registry/templates/registry/registry-deployment.yaml]: Ow, no reason. accidentally replaced. :)
i changed it.

File: charts/registry/templates/keycloak/init-script-configmap.yaml (hunk 10 above)
Comment (tunacicek) [reply charts/registry/templates/keycloak/init-script-configmap.yaml]: Done

File: charts/registry/templates/registry/registry-service.yaml (hunk 12 above)
Comment (tunacicek) [reply charts/registry/templates/registry/registry-service.yaml]: Done

**Comment by tunacicek** [comment]
@evegufy : Thank you for your feedback. I checkout the branch again and run it locally. In my case, the health checks (readiness, etc.) are running successfully. I can't reproduce the behavior. It may have taken a while for the service to start.

**Review by florianrusch-zf — APPROVED** [review]

**Comment by evegufy** [comment]
> @evegufy : Thank you for your feedback. I checkout the branch again and run it locally. In my case, the health checks (readiness, etc.) are running successfully. I can't reproduce the behavior. It may have taken a while for the service to start.

@tunacicek I tested a bit more, besides on my mac also on my linux machine.
I didn't observe the behaviour on linux and noticed then that on my mac the container never had enough time to start up entirely before the probes were causing a restart.
As a short term measure, I recommend increasing the initialDelaySeconds for readinessProbe and livenessProbe to 300 seconds as that solved the issue for me.
And you should look into providing the [image](https://hub.docker.com/r/tractusx/sldt-digital-twin-registry/tags) also for a arm64 cpu architecture, for better performance on newer macs.

**Comment by tunacicek** [comment]
> > @evegufy : Thank you for your feedback. I checkout the branch again and run it locally. In my case, the health checks (readiness, etc.) are running successfully. I can't reproduce the behavior. It may have taken a while for the service to start.
> 
> @tunacicek I tested a bit more, besides on my mac also on my linux machine. I didn't observe the behaviour on linux and noticed then that on my mac the container never had enough time to start up entirely before the probes were causing a restart. As a short term measure, I recommend increasing the initialDelaySeconds for readinessProbe and livenessProbe to 300 seconds as that solved the issue for me. And you should look into providing the [image](https://hub.docker.com/r/tractusx/sldt-digital-twin-registry/tags) also for a arm64 cpu architecture, for better performance on newer macs.

@evegufy , 
for the image arm64 topic i will create a issue. 
I made the values ​​for readinessProbe and livenessProbe configurable.

**Review by evegufy — APPROVED** [review]
